common/models/user.js

```diff
@@ -592,18 +609,30 @@ module.exports = function(User) {
   User.hashPassword = function(plain) {
     this.validatePassword(plain);
     var salt = bcrypt.genSaltSync(this.settings.saltWorkFactor || SALT_WORK_FACTOR);
-    return bcrypt.hashSync(plain, salt);
+    var hash = bcrypt.hashSync(plain, salt);
+    if (/^\$2a\$/.test(hash)) {
+      // change the prefix to indicate that we have verified the plain password
+      // is at most 72 chars long
+      hash = '$72$'.concat(hash.slice(4));
```

Comment:
@raymondfeng see https://github.com/strongloop/loopback/pull/2580#issuecomment-238186402:

> I am concerned about the backwards-incompatibility introduced in this patch. Because we refuse passwords longer than 72 characters at login time, if there is any existing user with such a long password, they won't be able to login any more after the server is upgraded to the newer LoopBack version.

In https://github.com/strongloop/loopback/pull/2580/commits/bfe674ce2097592fb30cce94f3a87eb84f0ff936, I implemented my proposal:

> Add some sort of a flag to distinguish between password that were created after 72 char validation was added (and thus are 72 chars at most) and the legacy passwords that we created before. Then at login time, we can reject too long values for accounts with the flag set, but allow them for legacy accounts. When a user resets their password, the new entry will be validated for maxlen and created with the flag, thus their account will be "upgraded" to the new, more secure mode.

There is one important constraint here: we cannot add new properties (fields) to the User model, because that would break existing applications upgrading to a newer LoopBack version - the database schema of their servers will not contain the newly added field.

In which case we have to encode the flag in the existing user fields. Encoding it in the password field makes most sense to me.

I decided to modify the hash instead of prepending/appending values to the hash to prevent another possible source of a breaking changes: what if users created their database columns as at most 60 chars long, because they new bcrypt hashes are always 60 chars long?

Having wrote that, I am open to better suggestions on how to solve this problem, as long as they preserve backwards compatibility.